### PR TITLE
Bug 1139259 - [Settings] Refactor call_fdn_list.js r=Arthur

### PR DESCRIPTION
**Pull request by EragonJ**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=12d9a26f9a97)

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c294298d8f21)

File: apps/settings/elements/call_fdn_settings.html [hunk 1]
```diff
@@ -41,7 +41,6 @@ <h1 data-l10n-id="fdnSettings-header"></h1>
     </div>
 
     <script src="js/simcard_dialog.js"></script>
```
Comment (crh0716) [inline apps/settings/elements/call_fdn_settings.html]: It seems we should also remove this line?

File: apps/settings/js/modules/fdn_context.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    fdnContacts: null,
+
+    /**
+     * we can use this method to get wrapped contacts information.
+     *
+     * @type {Function}
+     * @param {Function} callback
+     */
+    getContacts: function(callback) {
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: It would be better to return a promise.

File: apps/settings/js/modules/fdn_context.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+define(function() {
+  'use strict';
+
+  var FdnContext = {
+    /**
+     * cached fdnContacts
+     *
+     * @type {Array}
+     */
+    fdnContacts: null,
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: Add a `_` prefix if it is not intended to be exposed.

File: apps/settings/js/panels/call_fdn_list/call_fdn_list.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
+/* global DsdsSettings */
```
Comment (crh0716) [inline apps/settings/js/panels/call_fdn_list/call_fdn_list.js]: Let's require DsdsSettings explicitly.

File: apps/settings/elements/call_fdn_settings.html (hunk 1 above)
Comment (EragonJ) [inline apps/settings/elements/call_fdn_settings.html]: Hmm it seems that we forgot to remove this line last time, let me remove it here.

File: apps/settings/js/modules/fdn_context.js (hunk 3 above)
Comment (EragonJ) [inline apps/settings/js/modules/fdn_context.js]: ok

File: apps/settings/js/panels/call_fdn_list/call_fdn_list.js (hunk 4 above)
Comment (EragonJ) [inline apps/settings/js/panels/call_fdn_list/call_fdn_list.js]: ok

File: apps/settings/js/modules/fdn_context.js (hunk 2 above)
Comment (EragonJ) [inline apps/settings/js/modules/fdn_context.js]: ok, i am fine to change this interface

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=fc6bf5ca86f3)

File: apps/settings/js/modules/fdn_context.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    _fdnContacts: null,
+
+    /**
+     * we can use this method to get wrapped contacts information.
+     *
+     * @type {Function}
+     * @return {Promise}
+     */
+    getContacts: function() {
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: Let's make the function specific by passing the icc object or index directly.

File: apps/settings/js/modules/fdn_context.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+define(function() {
+  'use strict';
+
+  var FdnContext = {
+    /**
+     * cached fdnContacts
+     *
+     * @type {Array}
+     */
+    _fdnContacts: null,
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: The cache should support multiple sims.

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=4ab3b9776d3f)

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=76487244598b)

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=12bbe46f0e4c)

File: apps/settings/js/modules/fdn_context.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
+/**
+ * FdnContext is a module that you can easily fetch out FDN releated info.
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: type: related

File: apps/settings/js/modules/fdn_context.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     *   - existing id + name and number: update a contact
+     *   - existing id + empty name and number: remove a contact
+     *
+     * @type {Function}
+     * @param {String} action
+     * @param {Number} options.cardIndex
+     * @param {Object} options.contact
+     * @return {Object}
+     */
+    getContactInfo: function(action, options) {
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: Let's use `createAction` per the offline discussion.

File: apps/settings/js/modules/fdn_context.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      switch (action) {
+        case 'add':
+          simContact.name = [contact.name];
+          simContact.tel = [{
+            value: contact.number
+          }];
+          break;
+        case 'edit':
+          simContact = this._fdnContacts[cardIndex][contact.id];
+          simContact.name[0] = contact.name;
```
Comment (crh0716) [inline apps/settings/js/modules/fdn_context.js]: This modifies the contact in the cache before we do the action. We should clone the object.

File: apps/settings/js/modules/fdn_context.js (hunk 7 above)
Comment (EragonJ) [inline apps/settings/js/modules/fdn_context.js]: okay

File: apps/settings/js/modules/fdn_context.js (hunk 8 above)
Comment (EragonJ) [inline apps/settings/js/modules/fdn_context.js]: okay !

File: apps/settings/js/modules/fdn_context.js (hunk 9 above)
Comment (EragonJ) [inline apps/settings/js/modules/fdn_context.js]: okay !

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e28a7ea96fa9)

**Comment by try-server-hook** [comment]
[EragonJ EragonJ (E.J.) (EragonJ)](https://github.com/EragonJ) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=2e57bcfcfba8)

**Pull request merged**